tachyon/zk/plonk/circuit/expressions/evaluator/graph_evaluator.h

```diff
@@ -0,0 +1,271 @@
+          }
+        } else {
+          ValueSource right_result = AddExpression(sum->right());
+          if (left_result.IsZeroConstant()) {
+            return right_result;
+          } else if (right_result.IsZeroConstant()) {
+            return left_result;
+          } else {
+            // NOTE(chokobole): I don't know why this is needed.
+            // See
+            // https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/evaluation.rs#L709-L713.
+            if (left_result <= right_result) {
+              return AddCalculation(
+                  Calculation::Add(left_result, right_result));
+            } else {
+              return AddCalculation(
+                  Calculation::Add(right_result, left_result));
+            }
+          }
+        }
+      }
+
+      case ExpressionType::kProduct: {
+        const ProductExpression<F>* product = input->ToProduct();
+        ValueSource left_result = AddExpression(product->left());
+        ValueSource right_result = AddExpression(product->right());
+        if (left_result.IsZeroConstant() || right_result.IsZeroConstant()) {
+          return ValueSource::ZeroConstant();
+        } else if (left_result.IsOneConstant()) {
+          return right_result;
+        } else if (right_result.IsOneConstant()) {
+          return left_result;
+        } else if (left_result.IsTwoConstant()) {
+          return AddCalculation(Calculation::Double(right_result));
+        } else if (right_result.IsTwoConstant()) {
+          return AddCalculation(Calculation::Double(left_result));
+        } else if (left_result == right_result) {
+          return AddCalculation(Calculation::Square(left_result));
+        } else {
+          // NOTE(chokobole): we can do |left_result < right_result|.
+          // But ValueSource only implements <= operator and original source
+          // code also compares like below, too. See
+          // https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/evaluation.rs#L732
+          if (left_result <= right_result) {
+            return AddCalculation(Calculation::Mul(left_result, right_result));
+          } else {
+            return AddCalculation(Calculation::Mul(right_result, left_result));
+          }
+        }
+      }
+
+      case ExpressionType::kScaled: {
+        const ScaledExpression<F>* scaled = input->ToScaled();
+        const F& scale = scaled->scale();
+        if (scale.IsZero()) {
+          return ValueSource::ZeroConstant();
+        } else if (scale.IsOne()) {
+          return AddExpression(scaled->expr());
+        } else {
+          ValueSource constant = AddConstant(scale);
+          ValueSource result = AddExpression(scaled->expr());
+          return AddCalculation(Calculation::Mul(result, constant));
+        }
+      }
+    }
+    NOTREACHED();
```

Comment:
Is there a criterion for choosing between `NOTREACHED()` and `base::AlwaysFalse<>()`